### PR TITLE
Add Gecko-based text boxes (fix BL-964)

### PR DESCRIPTION
**Pull request by StephenMcConnel**
These are used optionally in the ArtOfReadingChooser control.  If the Web based thumbnail viewer is used, then this gecko-based text box needs to be used as well, at least on Linux.

### CODE REVIEW

File: PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBox.cs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+
+				EnsureFocusedGeckoControlHasInputFocus();
+
+				if (ContainsFocus)
+				{
+					return;
+				}
+
+				bool result = _browser.Focus();
+			};
```
Comment (papeh) [inline PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBox.cs]: bool result is never used

File: PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBase.cs [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+			return langName;
+		}
+
+		/// <summary>
+		/// Set InputFocus to a WinForm controls using Mono winforms connection to the X11 server.
+		/// GeckoWebBrowser.RemoveInputFocus uses the Gecko/Gtk connection to the X11 server.
+		/// The undoes the call to _browser.SetInputFocus.
+		/// Call this method when a winform control has gained focus, and X11 Input focus is still on the Gecko control.
+		/// </summary>
+		protected static void MoveInputFocusBacktoAWinFormsControl()
```
Comment (papeh) [inline PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBase.cs]: To should be capitalized

File: PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBase.cs [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+				UserGotFocus.Invoke(this, null);
+			}
+		}
+
+		protected override void OnEnter(EventArgs e)
+		{
+			_entered = true;
+
+			// Only do this is the control's html has loaded.
+			if (_browserDocumentLoaded)
```
Comment (papeh) [inline PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBase.cs]: Do this only *if* . . .
(or omit the comment entirely, as the if statement is pretty self-explanatory

File: PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBase.cs [hunk 4]
```diff
@@ -0,0 +1,532 @@
...
+		protected override void OnLeave(EventArgs e)
+		{
+			if (_browser != null)
+			{
+				this._browser.RemoveInputFocus();
+			}
+			if ((_browser != null) && (_browser.WebBrowserFocus != null))
+			{
+				_browser.WebBrowserFocus.Deactivate();
+			}
```
Comment (papeh) [inline PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBase.cs]: The above if statements could be nested

File: PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/ProgressUtils.cs [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+		/// </summary>
+		/// <param name="action">the action to run on UI thread</param>
+		/// <param name="exclusively">avoid overlap with ExecuteOnSameThread</param>
+		private static void InvokeOnUIThread(ThreadStart action, bool exclusively)
+		{
+			if (action == null)
+				return;
+
+			// Run directly if no UI context or if already on UI thread
+			if (OnMainUiThread)
```
Comment (papeh) [inline PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/ProgressUtils.cs]: As mentioned in the comment here, this property checks more than just whether we're on the UI thread.  The property name and property comment should reflect this.

File: PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBox.cs (hunk 1 above)
Comment (StephenMcConnel) [inline PalasoUIWindowsForms.GeckoFxWebBrowserAdapter/GeckoBox.cs]: David probably had it there for debugging.

File: PalasoUIWindowsForms/Widgets/ITextInputBox.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+using System.Collections.Generic;
+using System.ComponentModel;
+//using System.Diagnostics;
+using System.Drawing;
+//using System.Reflection;
+using System.Windows.Forms;
+//using Palaso.Reporting;
+//using Palaso.UI.WindowsForms.Keyboarding;
+//using Palaso.UI.WindowsForms.Spelling;
+using Palaso.WritingSystems;
```
Comment (papeh) [inline PalasoUIWindowsForms/Widgets/ITextInputBox.cs]: any reason for these commented-out usings?

**Comment by papeh** [comment]
Don't see anything too terribly objectionable; approved as soon as `OnMainUiThread` is renamed to `OnMainUiThreadOrNoUiContext` or similar.

**Comment by StephenMcConnel** [comment]
I've responded to papeh's comments, and also fixed the code to work on Windows (a couple of minor tweaks from what I had).

**Comment by StephenMcConnel** [comment]
Since I've been updating the commit as I've gone along, I'm not sure that github makes it easy for you the way gerrit does.  But if I can figure anything out shortly, I'll let you know.

**Comment by papeh** [comment]
Looks like my comments were addressed, and I couldn't find an easy way to look at an inter-patch diff (despite much googling and clicking around--maybe if I had the original commit hash?).  Passes the 10-second test, so LGTM :+1:

**Pull request merged**